web/components/templates/rateLimit/RateLimitRuleModal.tsx

```diff
@@ -0,0 +1,382 @@
+import { useMutation, useQueryClient } from "@tanstack/react-query";
+import { useOrg } from "@/components/layout/org/organizationContext"; // Import useOrg
+import {
+  Dialog,
+  DialogContent,
+  DialogHeader,
+  DialogTitle,
+  DialogDescription,
+  DialogFooter,
+} from "@/components/ui/dialog";
+import {
+  Select,
+  SelectContent,
+  SelectItem,
+  SelectTrigger,
+  SelectValue,
+} from "@/components/ui/select";
+import { Label } from "@/components/ui/label";
+import { Input } from "@/components/ui/input";
+import { Button } from "@/components/ui/button";
+import { P } from "@/components/ui/typography";
+import { components } from "@/lib/clients/jawnTypes/private"; // Import generated types
+import { Result } from "@/packages/common/result"; // <-- Add Result import
+
+// Use generated types instead of manual definitions
+type CreateRateLimitPayload =
+  components["schemas"]["CreateRateLimitRuleParams"];
+type UpdateRateLimitPayload =
+  components["schemas"]["UpdateRateLimitRuleParams"];
+type RateLimitRuleView = components["schemas"]["RateLimitRuleView"];
+
+// Zod schema for client-side validation (mirrors backend, suitable for both)
+const RateLimitRuleClientSchema = z.object({
+  name: z.string().min(1, "Rule Name is required."),
+  quota: z.number().nonnegative("Quota must be a non-negative number."),
+  window_seconds: z
+    .number()
+    .nonnegative("Time Window must be a non-negative number."),
+  unit: z.enum(["request", "cents"]),
+  segment: z
+    .string()
+    .optional() // Keep optional
+    .refine(
+      (val) =>
+        val === undefined || // Allow undefined (global)
+        val === "user" ||
+        /^[a-zA-Z0-9_-]+$/.test(val || ""), // Allow user or valid key
+      {
+        message:
+          "Segment must be 'user', empty (global), or a valid property key (alphanumeric/hyphen/underscore).",
+      }
+    ),
+});
+
+interface RateLimitRuleModalProps {
+  open: boolean;
+  onOpenChange: (isOpen: boolean) => void;
+  onSuccess: (rule: RateLimitRuleView) => void;
+  rule?: RateLimitRuleView; // <-- Make rule prop optional
+}
+
+const RateLimitRuleModal = ({
+  open,
+  onOpenChange,
+  onSuccess,
+  rule, // <-- Destructure optional rule
+}: RateLimitRuleModalProps) => {
+  const queryClient = useQueryClient();
+  const org = useOrg();
+  const isEditMode = !!rule; // Determine mode based on rule prop
+
+  // State variables (from original Create modal)
+  const [name, setName] = useState("");
+  const [quota, setQuota] = useState("");
+  const [unit, setUnit] = useState<"request" | "cents">("request");
+  const [windowSeconds, setWindowSeconds] = useState("");
+  const [segmentType, setSegmentType] = useState<
+    "global" | "user" | "property"
+  >("global");
+  const [customPropertyKey, setCustomPropertyKey] = useState("");
+  const [error, setError] = useState<string | null>(null);
+
+  // Effect to initialize/reset state based on mode
+  useEffect(() => {
```

Comment:
im not a frontend person at all, but this feels wrong, should we just put this all into a rule state?